### DISCUSSION TITLE
Relaxing direct_plugin_exports/ publish_fun

### DISCUSSION THREAD

**Discussion by mths1** [vernemq/vernemq#2089] [Ideas]
Hello all

For a plugin I am working on, I tested the direct_plugin_exports. Currently, it is a little bit limited in what in can do (what you can publish). In particular, I would like to add at least user properties as I heavily depend on them. It is my understanding that the publish function works on the internal datamodel just as it would drop out of the mqtt v3 and mqtt v5 parser. 

So, it wouldn't do much harm if it is relaxed slightly as long as it is ensured that it doesn't end up in some illegal state (a state that wouldn't have been produced by mqtt v5 parsers).

I actually tested it already and it works nicely with MQTT v3 and MQTT v5 subscribers (of course, only the later one actually receives the user properties).

**Comment by ioolkos** [comment]
`direct_plugin_exports` allow a plugin to "implement" (should rather say: "apply") a generic Register/Subscribe/Publish function in a plugin.
Normal plugin hooks on the contrary allow you to react to Register/Subcribe/Publish events externally, which is usually what plugins are interested in.
So, the domain is different; the former is directed towards the broker. This is just for context.

If it's done safely, I will support extending the `direct_plugin_export` Funs. @mths1, Maybe you can provide a concrete example? The MQTT v3/v5 point is valid too; for instance, the `vmq_bridge` plugin (which is a user of `direct_plugin_exports`) does not support v5 properties yet.

**Reply by mths1** [reply]
It would enable all sorts of bridging scenarios (MQTTv5 Bridge, REST, Kafka, PubSub, RPC...). I have seen that vmq_bridge does not support MQTTv5 yet. Is gen_mqtt_client v5 ready? In that case, by extending direct_plugin_export supporting v5 would be possible without too much effort...

BTW: Is it possible for a plugin to trigger an authentication for a given client (given that it knows client_id and the necessary secrets)?

**Reply by ioolkos** [reply]
`gen_mqtt_client` does not support v5. The only need ever to have a client implementation was the bridge, so it's kind of minimal. Shouldn't call it simple, though, it's quite a crazy FSM :)

For authentication, maybe look at https://github.com/vernemq/vernemq_dev/blob/master/src/vernemq_dev_api.erl where there's a function that reauthorizes subscriptions (only). For full authentication (let's say pushing the Client through `auth_on_register` again somehow) the question to think about is the disconnect. I mean, if you end up disconnecting the client, then you could as well disconnect it right away to trigger authentication. But maybe this is a topic I need to look into more in depth.
There's also this enhanced authentication stuff in v5 that nobody ever uses (in my experience so far). We might need to review that too at some point.

**Reply by mths1** [reply]
As far as the authentication goes I now used:

```
   case vmq_plugin:all_till_ok(auth_on_register, HookArgs) of
        ok ->
            [...]

```
which seems to be all I need and should be more or less side-effect free (might have an impact on caches...). Of course, that is kind of a hack...
